include/nanobind/eigen/dense.h

```diff
@@ -241,6 +249,107 @@ struct type_caster<Eigen::Ref<T, Options, StrideType>> {
+
+    using ScalarCaster = make_caster<tensor<Scalar>>;
+    using StorageIndexCaster = make_caster<tensor<StorageIndex>>;
+    using IndexCaster = make_caster<Index>;
+
+    NB_TYPE_CASTER(T, const_name<rowMajor>("scipy.sparse.csr_matrix[",
+					   "scipy.sparse.csc_matrix[")
+		   + make_caster<Scalar>::Name + const_name("]"));
+
+    ScalarCaster scalar_caster;
+    StorageIndexCaster storageindex_caster;
+    IndexCaster index_caster;
+
+    bool from_python(handle src, uint8_t flags, cleanup_list *cleanup) noexcept {
+	if (!src) {
+	    return false;
+	}
+
+	auto obj = borrow<object>(src);
+	object sparse_module = module_::import_("scipy.sparse");
+	object matrix_type = sparse_module.attr(rowMajor ? "csr_matrix" : "csc_matrix");
+
+	if (!obj.type().is(matrix_type)) {
+            try {
+                obj = matrix_type(obj);
+            } catch (const python_error &) {
+                return false;
+            }
+        }
+
+	if (!scalar_caster.from_python(obj.attr("data"), flags, cleanup)) {
+	    return false;
+	}
+        auto values = scalar_caster.value;
+
+	if (!storageindex_caster.from_python(obj.attr("indices"), flags, cleanup)) {
+	    return false;
+	}
+        auto innerIndices = storageindex_caster.value;
+
+	if (!scalar_caster.from_python(obj.attr("indptr"), flags, cleanup)) {
+	    return false;
+	}
+        auto outerIndices = storageindex_caster.value;
+
+	// FIXME: Write with caster.from_python
+	auto shape = cast<std::tuple<Index, Index>>(obj.attr("shape"));
+
+	if (!index_caster.from_python(obj.attr("nnz"), flags, cleanup)) {
+	    return false;
+	}
+        auto nnz = index_caster.value;
+
+        value = EigenMapSparseMatrix<Scalar,
+                                     T::Flags &(Eigen::RowMajor | Eigen::ColMajor),
+                                     StorageIndex>(std::get<0>(shape),
+						   std::get<1>(shape),
+                                                   std::move(nnz),
+                                                   outerIndices.data(),
+                                                   innerIndices.data(),
+                                                   values.data());
+
+        return true;
+    }
+
+    static handle from_cpp(T &&v, rv_policy policy, cleanup_list *cleanup) noexcept {
+        if (policy == rv_policy::automatic ||
+            policy == rv_policy::automatic_reference)
+            policy = rv_policy::move;
+
+        return from_cpp((const T &) v, policy, cleanup);
+    }
+
+    static handle from_cpp(const T &v, rv_policy policy, cleanup_list *cleanup) noexcept {
+	// FIXME: Looks like UB. Is there anything that can be done about this?
+	T & src = const_cast<T &>(v);
+	src.makeCompressed();
+
+	object sparse_module = module_::import_("scipy.sparse");
+	object matrix_type = sparse_module.attr(rowMajor ? "csr_matrix" : "csc_matrix");
+
+	// FIXME: These casts are ugly! Is there another way to write this?
+	const size_t data_shape[] = { static_cast<size_t>(v.nonZeros()) };
```

Comment:
using `static_cast` for an integer conversion seems quite overkill.

Reply:
Is there something else that bothers you here?

Reply:
The cast is needed to narrow the results of the calls to `size_t` which is required for the `nb::tensor` constructor. Leaving out the cast generates a compiler warning in GCC and an error in Clang:
```
nanobind/include/nanobind/eigen/sparse.h:115:32: error: non-constant-expression cannot be narrowed from type 'Eigen::Index' (aka 'long') to 'size_t' (aka 'unsigned long') in initializer list [-Wc++11-narrowing]
        const size_t data_shape[] = { v.nonZeros() };
                                      ^~~~~~~~~~~~
```
Why do you think `static_cast` is overkill? What's the alternative?

Reply:
Just a c-style cast. I think static_cast is great when you have polymorphic classes and other complex aspects of C++ and want to be extra-clear about what kind of cast is acceptable. But I don't see the advantage for simple integer casts, and it just makes the code less readable IMO.

Reply:
But according to https://eel.is/c++draft/expr.cast `(T)` will just try in order `const_cast<T>`, `static_cast<T>`, `reinterpret_cast<T>`. See also https://stackoverflow.com/questions/103512/why-use-static-castintx-instead-of-intx. But I'm happy to do C-style casts as well if you wish.

Reply:
Right. In those stackoverflow threads, you can see that there are footguns involving C++ classes, implicit conversions involving them, and polymorphism. None of which applies to a simple integer cast. For me, using static_cast even in such simple situations is an anti-pattern that makes code ugly to read. (This is somewhat subjective, I know that other people may disagree on what a good default policy is).